var/spack/repos/builtin/packages/py-cupy/package.py

```diff
@@ -31,18 +35,69 @@ class PyCupy(PythonPackage, CudaPackage):
     depends_on("py-fastrlock@0.5:", type=("build", "run"))
     depends_on("py-numpy@1.20:1.25", when="@:11", type=("build", "run"))
     depends_on("py-numpy@1.20:1.26", when="@12:", type=("build", "run"))
-    depends_on("py-scipy@1.6:1.12", type=("build", "run"))
-    depends_on("cuda@:11.9", when="@:11")
-    depends_on("cuda@:12.1", when="@12:")
-    depends_on("nccl")
-    depends_on("cudnn")
-    depends_on("cutensor")
 
-    conflicts("~cuda")
+    depends_on("py-scipy@1.6:1.12", when="+all", type=("build", "run"))
+    depends_on("py-cython@0.29.22:2", when="+all", type=("build", "run"))
+    depends_on("py-optuna@2:", when="+all", type=("build", "run"))
+
+    # Based on https://github.com/cupy/cupy/releases
+    depends_on("cuda@:11.9", when="@:11 +cuda")
+    depends_on("cuda@:12.1", when="@12: +cuda")
+
+    for a in CudaPackage.cuda_arch_values:
+        depends_on("nccl +cuda cuda_arch={0}".format(a), when="+cuda cuda_arch={0}".format(a))
```

Comment:
I've never been a big fan of variant forwarding. Personally, I would rather leave this up to the user/concretizer instead of the packager. Users can now specify `cuda_arch==...` and it will automatically be propagated to all dependencies.

Reply:
Variant forwarding via `cuda_arch==` does not work with `nccl` here:
```
$> cat spack.yaml
...
  specs:
  - py-cupy +all +cuda cuda_arch==80

$> spack -e . concretize -f | tee concretize.log
==> Starting concretization
==> Error: concretization failed for the following reasons:

   1. Must specify CUDA compute capabilities of your GPU, see https://developer.nvidia.com/cuda-gpus

real	0m31.969s
user	0m31.004s
sys	0m0.947s
```

Reply:
* https://github.com/spack/spack/issues/39168